apps/site/assets/ts/stop/components/DepartureTimes.tsx

```diff
@@ -1,35 +1,333 @@
 import React, { ReactElement } from "react";
 import usePredictionsChannel from "../../hooks/usePredictionsChannel";
 import { isACommuterRailRoute } from "../../models/route";
-import { DirectionId, Route, Stop } from "../../__v3api";
+import {
+  DirectionId,
+  PredictionForStop,
+  Route,
+  ScheduleForStop,
+  Stop
+} from "../../__v3api";
 import renderFa from "../../helpers/render-fa";
-import { formatDepartureTime } from "../../helpers/date";
+import { formatToBostonTime } from "../../helpers/date";
+import { filter, find, groupBy, slice } from "lodash";
+import {
+  differenceInSeconds,
+  isSameDay,
+  secondsInDay,
+  secondsInHour,
+  secondsInMinute
+} from "date-fns";
+import realtimeIcon from "../../../static/images/icon-realtime-tracking.svg";
+import SVGIcon from "../../helpers/render-svg";
 
 interface DepartureTimesProps {
   route: Route;
   stop: Stop;
   directionId: DirectionId;
+  schedulesForDirection: ScheduleForStop[];
 }
 
+// This interface is used to tell the front end
+// how to display the ScheduleInfoModel data
+interface DisplayTimeConfig {
+  isPrediction?: boolean;
+  displayString: string;
+  trackName?: string | null;
+  isTomorrow?: boolean;
+  isBolded?: boolean;
+  isStikethrough?: boolean;
+}
+
+// This interface is used to combine the information from the schedules
+// and predictions into one model
+interface ScheduleInfoModel {
+  time: Date;
+  isCancelled?: boolean;
+  isDelayed?: boolean;
+  track?: string | null;
+}
+
+const isCancelled = (prediction: PredictionForStop): boolean => {
+  return prediction.schedule_relationship === "cancelled";
+};
+
+// Finds the corresponding schedule to the prediction and compares the times
+// If the prediction an schedule are more than 60 seconds apart it is delayed
+const isDelayed = (
+  prediction: PredictionForStop,
+  schedules: ScheduleForStop[],
+  delay: number = 60
+): boolean => {
+  const schedule = find(
+    schedules,
+    (sch: ScheduleForStop) => sch.trip.id === prediction.trip.id
+  );
+
+  return (
+    !!schedule &&
+    // Is the prediction meaningfully after the schedule
+    differenceInSeconds(prediction.time, schedule.time) > delay
+  );
+};
+
+// Assuming there are predictions
+// This function will return an array containing at least one predcition.
+// It will return 2 predictions if there are more than 2 predictions that are not cancelled,
+// If the first prediction is cancelled, the 2nd prediction will not be cancelled
+const filterPredictions = (
+  predictions: PredictionForStop[]
+): Array<PredictionForStop | undefined> => {
```

Comment:
As written right now it will return either `[prediction, prediction]` or `[prediction, undefined]`
The only time the 2nd case would happen is if all the predictions we have are cancelled.

I am thinking about rewriting it  so it should return `[prediction, prediction]` or `[prediction]`.  Though down stream it will still get expanded to `[prediction, undefined]` when used in the `getNextTwoTimes` function